{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "from nltk import pos_tag\n",
    "from nltk.tokenize import word_tokenize\n",
    "\n",
    "# USING_PORTION = slice(None, 80) # When training\n",
    "# USING_PORTION = slice(80, None) # When testing\n",
    "USING_PORTION = slice(0, 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================ LOADING DATASET ================\n",
      "Poly- and perfluorinated compounds activate human pregnane X receptor.\n",
      "Neurotransmitters activate T-cells and elicit crucial functions via neurotransmitter receptors\n",
      "Pericytes activate complement in fibrosis\n",
      "These molecules activate different signaling pathways and produce different physiological responses which can be taken advantage of for sepsis modeling.\n",
      "Monoamines activate neuropeptide signaling cascades to modulate nociception in C. elegans\n"
     ]
    }
   ],
   "source": [
    "dataset = []\n",
    "with open('./dataset/dataset.txt', encoding='utf-8') as data_file:\n",
    "    dataset = data_file.readlines()\n",
    "\n",
    "dataset = [data.strip() for data in dataset]\n",
    "\n",
    "print(\"================ LOADING DATASET ================\")\n",
    "print('\\n'.join(dataset[:5]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "================ TAGGING DATASET ================\n",
      "[('Poly-', 'NNP'), ('and', 'CC'), ('perfluorinated', 'VBD'), ('compounds', 'NNS'), ('activate', 'V'), ('human', 'JJ'), ('pregnane', 'NN'), ('X', 'NNP'), ('receptor', 'NN'), ('.', '.')]\n",
      "[('Neurotransmitters', 'NNS'), ('activate', 'V'), ('T-cells', 'NNS'), ('and', 'CC'), ('elicit', 'JJ'), ('crucial', 'JJ'), ('functions', 'NNS'), ('via', 'IN'), ('neurotransmitter', 'NN'), ('receptors', 'NNS')]\n",
      "[('Pericytes', 'NNS'), ('activate', 'V'), ('complement', 'NN'), ('in', 'IN'), ('fibrosis', 'NN')]\n",
      "[('These', 'DT'), ('molecules', 'NNS'), ('activate', 'V'), ('different', 'JJ'), ('signaling', 'NN'), ('pathways', 'NNS'), ('and', 'CC'), ('produce', 'VB'), ('different', 'JJ'), ('physiological', 'JJ'), ('responses', 'NNS'), ('which', 'WDT'), ('can', 'MD'), ('be', 'VB'), ('taken', 'VBN'), ('advantage', 'NN'), ('of', 'IN'), ('for', 'IN'), ('sepsis', 'NN'), ('modeling', 'NN'), ('.', '.')]\n",
      "[('Monoamines', 'NNS'), ('activate', 'V'), ('neuropeptide', 'IN'), ('signaling', 'VBG'), ('cascades', 'NNS'), ('to', 'TO'), ('modulate', 'VB'), ('nociception', 'NN'), ('in', 'IN'), ('C.', 'NNP'), ('elegans', 'NNS')]\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "# To mitigate pos_tag wrongly tags\n",
    "#     If these words appear, force tag it as 'V'\n",
    "verbs = [\n",
    "    'activate', 'inhibit', 'bind',\n",
    "    'accelerate', 'augment', 'induce', 'stimulate', 'require', 'up-regulate',\n",
    "    'abolish', 'block', 'down-regulate', 'prevent'\n",
    "]\n",
    "\n",
    "remove_paren = re.compile(r\"\\(.*?\\)\")\n",
    "dataset_preprocessed = [\n",
    "    remove_paren.sub('', sentence)\n",
    "    for sentence in dataset\n",
    "]\n",
    "\n",
    "dataset_tagged = [\n",
    "    [\n",
    "        (word, tag) if word not in verbs else (word, 'V')\n",
    "        for word, tag in pos_tag(word_tokenize(sentence))\n",
    "    ]\n",
    "    for sentence in dataset_preprocessed\n",
    "]\n",
    "\n",
    "print(\"================ TAGGING DATASET ================\")\n",
    "print('\\n'.join([ str(dataset) for dataset in dataset_tagged[:5] ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Parsed 73 sents, Unable to parse 27 sents\n"
     ]
    }
   ],
   "source": [
    "from nltk import CFG\n",
    "from nltk.parse import ChartParser\n",
    "from nltk.grammar import Nonterminal, Production\n",
    "\n",
    "grammar = CFG.fromstring('\\n'.join([\n",
    "    'S -> NP VP | NP VP NP | S C S | S \".\" | VP NP | WDT S | EX S | S RB | S PART | IN S',\n",
    "    'NP -> N | DT NP | J NP | NP NP | NP MOD | NP C NP | NP R | R NP | PART | NP POS | WRONG_NP',\n",
    "    'VP -> V | R VP | MD VP | VP PART | VP C VP | VP MOD',\n",
    "    'MOD -> IN NP | IN IN NP | TO VP | TO NP',\n",
    "    'PART -> VBN | VBG',\n",
    "    'C -> CC | \",\"',\n",
    "    'N -> NN | NNP | NNS | FW | CD | PRP | WRONG_N',\n",
    "    'R -> RB | RBR | RBS | RP | R C R',\n",
    "    'V -> VB | VBD | VBP | VBZ',\n",
    "    'J -> JJ | JJR | JJS | J C J',\n",
    "    'WRONG_NP -> JJ | R V | NNP VBZ'\n",
    "])).productions()\n",
    "\n",
    "\n",
    "# Append pos_tagged results into originally defined grammar\n",
    "def generate_sent_grammar(sent):\n",
    "    sent_grammar = grammar[:]\n",
    "    sent_words = list([word for word, tag in sent])\n",
    "    \n",
    "    for word, tag in sent:\n",
    "        sent_grammar.append(Production(\n",
    "            Nonterminal(tag), (word, )\n",
    "        ))\n",
    "    \n",
    "    return sent_grammar, sent_words\n",
    "\n",
    "    \n",
    "non_sents = {}\n",
    "parsed_sents = {}\n",
    "# Parse dataset\n",
    "for (idx, sent) in enumerate(dataset_tagged[USING_PORTION]):\n",
    "    sent_grammar, sent_words = generate_sent_grammar(sent)\n",
    "    \n",
    "    parser = ChartParser(CFG(Nonterminal('S'), sent_grammar))\n",
    "    tree = parser.parse_one(sent_words)\n",
    "    \n",
    "    if tree is None:\n",
    "        non_sents[idx] = sent\n",
    "    \n",
    "    else:\n",
    "        parsed_sents[idx] = tree\n",
    "\n",
    "print(\"Parsed %d sents, Unable to parse %d sents\" % (len(parsed_sents), len(non_sents)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 [{\"X\": \"Poly- and perfluorinated compounds\", \"V\": \"activate\", \"Y\": \"human pregnane X receptor .\"}]\n",
      "1 [{\"X\": \"Neurotransmitters\", \"V\": \"activate\", \"Y\": \"T-cells\"}, {\"X\": \"Neurotransmitters\", \"V\": \"activate\", \"Y\": \"functions receptors\"}]\n",
      "2 [{\"X\": \"Pericytes\", \"V\": \"activate\", \"Y\": \"complement\"}]\n",
      "3 [{\"X\": \"molecules\", \"V\": \"activate\", \"Y\": \"different signaling pathways and produce different physiological responses which can be taken advantage of for sepsis modeling .\"}]\n",
      "4 [{\"X\": \"Monoamines\", \"V\": \"activate\", \"Y\": \"elegans\"}]\n",
      "5 [{\"X\": \"bacteria\", \"V\": \"activate\", \"Y\": \"interferon production\"}]\n",
      "6 [{\"X\": \"Salient sounds\", \"V\": \"activate\", \"Y\": \"human visual cortex automatically .\"}]\n",
      "7 [{\"X\": \"Bisphenol A and its analogues\", \"V\": \"activate\", \"Y\": \"human pregnane X receptor .\"}]\n",
      "8 [{\"X\": \"cancer vaccination\", \"V\": \"activate\", \"Y\": \"T-cell responses\"}, {\"X\": \"cells\", \"V\": \"activate\", \"Y\": \"T-cell responses\"}]\n",
      "9 [{\"X\": \"Because the subthreshold sequences\", \"V\": \"activate\", \"Y\": \"the retina at stimulation rates that would typically induce strong fading , such retinal 'tickling ' has the potential to minimize the fading problem\"}]\n",
      "10 [{\"X\": \"Instead of achieving reward through adaptive behaviors , drugs of abuse directly\", \"V\": \"activate\", \"Y\": \"the reward pathways .\"}]\n",
      "11 [{\"X\": \"Cannabinoids\", \"V\": \"activate\", \"Y\": \"signaling Key C. elegans behaviors\"}]\n",
      "12 [{\"X\": \"Integrins\", \"V\": \"activate\", \"Y\": \"transcription factor-kappaB\"}, {\"X\": \"cytokines\", \"V\": \"activate\", \"Y\": \"transcription factor-kappaB\"}]\n",
      "13 [{\"X\": \"Enhancer elements\", \"V\": \"activate\", \"Y\": \"3 ' splice site 2\"}]\n",
      "14 [{\"X\": \"actions\", \"V\": \"activate\", \"Y\": \"sulcus\"}, {\"X\": \"actions\", \"V\": \"activate\", \"Y\": \"fusiform gyrus perceptual cues\"}]\n",
      "15 [{\"X\": \"Muscarinic acetylcholine receptors\", \"V\": \"activate\", \"Y\": \"acetylcholinesterase gene promoter\"}]\n",
      "16 [{\"X\": \"Anti-cyclic citrullinated peptide antibodies from rheumatoid arthritis patients\", \"V\": \"activate\", \"Y\": \"complement via both the classical and alternative pathways .\"}]\n",
      "17 [{\"X\": \"ions\", \"V\": \"activate\", \"Y\": \"nitroalkenes bonding\"}]\n",
      "18 [{\"X\": \"Anti-capsular antibodies\", \"V\": \"activate\", \"Y\": \"killing of Escherichia coli O8 : K87 by the alternate complement pathway in porcine serum .\"}]\n",
      "19 [{\"X\": \"compounds\", \"V\": \"activate\", \"Y\": \"TRPA1 ion channels\"}]\n",
      "20 [{\"X\": \"Bacterial polysaccharides\", \"V\": \"inhibit\", \"Y\": \"hyperglycemia\"}]\n",
      "21 [{\"X\": \"Iridoids nucleus\", \"V\": \"inhibit\", \"Y\": \"inflammation\"}, {\"X\": \"Iridoids nucleus\", \"V\": \"inhibit\", \"Y\": \"stress NF-\\u03baB pathway\"}]\n",
      "22 [{\"X\": \"pseudomonads\", \"V\": \"inhibit\", \"Y\": \"Oomycete plant pathogens max\"}]\n",
      "23 [{\"X\": \"Phenformin,metformin\", \"V\": \"inhibit\", \"Y\": \"growth and migration of LN229 glioma cells in vitro and in vivo .\"}]\n",
      "24 [{\"X\": \"Bone marrow-derived mesenchymal stem cells\", \"V\": \"inhibit\", \"Y\": \"T follicular helper cell mice\"}]\n",
      "25 [{\"X\": \"Bufadienolides\", \"V\": \"inhibit\", \"Y\": \"cell proliferation\"}, {\"X\": \"Bufadienolides\", \"V\": \"inhibit\", \"Y\": \"arrest cells phase\"}]\n",
      "26 [{\"X\": \"Flavonoids\", \"V\": \"inhibit\", \"Y\": \"enzymes\"}, {\"X\": \"Flavonoids\", \"V\": \"inhibit\", \"Y\": \"production\"}]\n",
      "27 [{\"X\": \"Procyanidins\", \"V\": \"inhibit\", \"Y\": \"angiogenesis extracellular matrix\"}]\n",
      "28 [{\"X\": \"hydrophobic aggregates\", \"V\": \"inhibit\", \"Y\": \"formation\"}]\n",
      "29 [{\"X\": \"Cytotropic heterogeneous molecular lipids\", \"V\": \"inhibit\", \"Y\": \"growth cells\"}]\n",
      "30 [{\"X\": \"Oxathiazolones\", \"V\": \"inhibit\", \"Y\": \"immunoproteasome\"}]\n",
      "31 [{\"X\": \"Largazole and its derivatives selectively\", \"V\": \"inhibit\", \"Y\": \"ubiquitin activating enzyme .\"}]\n",
      "32 [{\"X\": \"Jun proteins\", \"V\": \"inhibit\", \"Y\": \"autophagy\"}]\n",
      "33 [{\"X\": \"Axl-148b chimeric aptamers\", \"V\": \"inhibit\", \"Y\": \"cancer\"}, {\"X\": \"Axl-148b chimeric aptamers\", \"V\": \"inhibit\", \"Y\": \"melanoma progression\"}]\n",
      "34 [{\"X\": \"receptor 2 agonists\", \"V\": \"inhibit\", \"Y\": \"fibrocyte differentiation\"}]\n",
      "35 [{\"X\": \"Ebola virus VP24 proteins\", \"V\": \"inhibit\", \"Y\": \"interaction karyopherin alpha proteins STAT1\"}]\n",
      "36 [{\"X\": \"Hawthorn extracts\", \"V\": \"inhibit\", \"Y\": \"LDL oxidation\"}]\n",
      "37 [{\"X\": \"membranes\", \"V\": \"inhibit\", \"Y\": \"prostaglandin\"}, {\"X\": \"membranes\", \"V\": \"inhibit\", \"Y\": \"contractions\"}]\n",
      "38 [{\"X\": \"monocytes\", \"V\": \"inhibit\", \"Y\": \"multiplication pneumophila\"}]\n",
      "39 [{\"X\": \"Heparin\", \"V\": \"inhibit\", \"Y\": \"zymosan\"}, {\"X\": \"Heparin\", \"V\": \"inhibit\", \"Y\": \"cobra venom factor activation\"}, {\"X\": \"derivatized heparin\", \"V\": \"inhibit\", \"Y\": \"zymosan\"}, {\"X\": \"derivatized heparin\", \"V\": \"inhibit\", \"Y\": \"cobra venom factor activation\"}]\n",
      "40 [{\"X\": \"In contrast ,\", \"V\": \"amphetamine\", \"Y\": \"only binds to the N-terminus .\"}]\n",
      "41 [{\"X\": \"Type I inhibitors\", \"V\": \"bind\", \"Y\": \"to the active protein kinase conformation .\"}]\n",
      "42 [{\"X\": \"TALE proteins\", \"V\": \"bind\", \"Y\": \"chromatin\"}]\n",
      "43 [{\"X\": \"exon 1 amyloid fibrils\", \"V\": \"bind\", \"Y\": \"laterally to the cellular membrane .\"}]\n",
      "44 [{\"X\": \"Catecholamines\", \"V\": \"bind\", \"Y\": \"to enkephalins , morphiceptin , and morphine .\"}]\n",
      "45 [{\"X\": \"Acanthamoebae\", \"V\": \"bind\", \"Y\": \"epithelium\"}]\n",
      "46 [{\"X\": \"Bordetella pertussis fimbriae\", \"V\": \"bind\", \"Y\": \"monocytes FimD\"}]\n",
      "47 [{\"X\": \"pathogenic bacteria\", \"V\": \"bind\", \"Y\": \"specifically to the carbohydrate sequence GalNAc beta 1-4Gal found in some glycolipids .\"}]\n",
      "48 [{\"X\": \"TTAGGG repeat factors 1\", \"V\": \"bind\", \"Y\": \"subset\"}, {\"X\": \"TTAGGG repeat factors 1\", \"V\": \"bind\", \"Y\": \"repeats\"}, {\"X\": \"2\", \"V\": \"bind\", \"Y\": \"subset\"}, {\"X\": \"2\", \"V\": \"bind\", \"Y\": \"repeats\"}]\n",
      "49 [{\"X\": \"membrane-bound\", \"V\": \"bind\", \"Y\": \"bacteria\"}, {\"X\": \"forms\", \"V\": \"bind\", \"Y\": \"bacteria\"}]\n",
      "50 [{\"X\": \"Anti-ribosomal and ' P-peptide'-specific autoantibodies\", \"V\": \"bind\", \"Y\": \"to T lymphocytes .\"}]\n",
      "51 [{\"X\": \"CD3+ thymocytes\", \"V\": \"bind\", \"Y\": \"cells\"}]\n",
      "52 [{\"X\": \"Sulfoglucuronyl glycolipids\", \"V\": \"bind\", \"Y\": \"laminin\"}]\n",
      "53 [{\"X\": \"Hyaluronate binding proteins\", \"V\": \"bind\", \"Y\": \"to fibronectin , laminin and collagen .\"}]\n",
      "54 [{\"X\": \"Microvesicles\", \"V\": \"bind\", \"Y\": \"soluble fibrinogen , adhere to immobilized fibrinogen and coaggregate with platelets .\"}]\n",
      "55 [{\"X\": \"Falciparum malaria-infected erythrocytes\", \"V\": \"bind\", \"Y\": \"cells\"}]\n",
      "56 [{\"X\": \"monomer\", \"V\": \"bind\", \"Y\": \"TLR-2\"}, {\"X\": \"peptides\", \"V\": \"bind\", \"Y\": \"TLR-2\"}]\n",
      "57 [{\"X\": \"Scorpion toxins affecting sodium current inactivation\", \"V\": \"bind\", \"Y\": \"receptor sites channels\"}]\n",
      "58 [{\"X\": \"fingers\", \"V\": \"bind\", \"Y\": \"DNA\"}]\n",
      "59 [{\"X\": \"Enzymes\", \"V\": \"bind\", \"Y\": \"substrates at key locations in their structure called active sites\"}]\n",
      "60 [{\"X\": \"Obesity\", \"V\": \"accelerate\", \"Y\": \"aging process\"}]\n",
      "61 [{\"X\": \"Antioxidants\", \"V\": \"accelerate\", \"Y\": \"lung cancer progression\"}]\n",
      "62 [{\"X\": \"Grooves\", \"V\": \"accelerate\", \"Y\": \"dew shedding\"}]\n",
      "63 [{\"X\": \"proton bunches\", \"V\": \"accelerate\", \"Y\": \"electrons\"}]\n",
      "64 [{\"X\": \"Rat\", \"V\": \"accelerate\", \"Y\": \"osteogenesis\"}, {\"X\": \"exosomes\", \"V\": \"accelerate\", \"Y\": \"osteogenesis\"}]\n",
      "65 [{\"X\": \"TFS\", \"V\": \"accelerate\", \"Y\": \"transcription\"}, {\"X\": \"Spt4/5\", \"V\": \"accelerate\", \"Y\": \"transcription\"}]\n",
      "66 [{\"X\": \"Eosinophils\", \"V\": \"accelerate\", \"Y\": \"pathogenesis of psoriasis by supporting an inflammatory milieu that promotes neutrophil infiltration .\"}]\n",
      "67 [{\"X\": \"Lipid membranes\", \"V\": \"accelerate\", \"Y\": \"formation model amyloidosis\"}]\n",
      "68 [{\"X\": \"Homocysteine\", \"V\": \"accelerate\", \"Y\": \"skin\"}]\n",
      "69 [{\"X\": \"NLRP3 inflammasome\", \"V\": \"accelerate\", \"Y\": \"mediated pathology hepatitis\"}, {\"X\": \"IL-1\\u03b2\", \"V\": \"accelerate\", \"Y\": \"mediated pathology hepatitis\"}]\n",
      "70 [{\"X\": \"glucosamine fibers\", \"V\": \"accelerate\", \"Y\": \"hemostasis treated drugs\"}]\n",
      "71 [{\"X\": \"administered adipose-derived stem cells\", \"V\": \"accelerate\", \"Y\": \"healing\"}, {\"X\": \"administered adipose-derived stem cells\", \"V\": \"accelerate\", \"Y\": \"vasculogenesis\"}]\n",
      "72 [{\"X\": \"Cumulus cells\", \"V\": \"accelerate\", \"Y\": \"factor\"}]\n",
      "73 [{\"X\": \"Aluminium salts\", \"V\": \"accelerate\", \"Y\": \"peroxidation\"}]\n",
      "74 [{\"X\": \"foundations\", \"V\": \"accelerate\", \"Y\": \"change\"}, {\"X\": \"foundations\", \"V\": \"accelerate\", \"Y\": \"competence\"}]\n",
      "75 [{\"X\": \"Gangliosides\", \"V\": \"accelerate\", \"Y\": \"rat neonatal learning\"}, {\"X\": \"Gangliosides\", \"V\": \"accelerate\", \"Y\": \"levels acetylcholinesterases\"}]\n",
      "76 [{\"X\": \"Semax\", \"V\": \"accelerate\", \"Y\": \"healing\"}, {\"X\": \"glyproline peptides\", \"V\": \"accelerate\", \"Y\": \"healing\"}]\n",
      "77 [{\"X\": \"Faulty engines cells\", \"V\": \"accelerate\", \"Y\": \"ageing\"}, {\"X\": \"Faulty engines cells\", \"V\": \"accelerate\", \"Y\": \"disease\"}]\n",
      "78 [{\"X\": \"MIF-1\", \"V\": \"accelerate\", \"Y\": \"neuromotor\"}, {\"X\": \"MIF-1\", \"V\": \"accelerate\", \"Y\": \"EEG\"}, {\"X\": \"MIF-1\", \"V\": \"accelerate\", \"Y\": \"development\"}]\n",
      "79 [{\"X\": \"Distributed representations\", \"V\": \"accelerate\", \"Y\": \"evolution\"}]\n",
      "80 [{\"X\": \"Azeloglycine\", \"V\": \"down-regulate\", \"Y\": \"mediators\"}, {\"X\": \"nicotinamide\", \"V\": \"down-regulate\", \"Y\": \"mediators\"}]\n",
      "81 [{\"X\": \"miR-214\", \"V\": \"down-regulate\", \"Y\": \"Necl-2/CADM1\"}, {\"X\": \"hypoxia\", \"V\": \"down-regulate\", \"Y\": \"Necl-2/CADM1\"}]\n",
      "82 [{\"X\": \"let-7\", \"V\": \"down-regulate\", \"Y\": \"axon-guidance genes Ntn1\"}, {\"X\": \"let-7\", \"V\": \"down-regulate\", \"Y\": \"Dcc nerve regeneration\"}, {\"X\": \"miR-9\", \"V\": \"down-regulate\", \"Y\": \"axon-guidance genes Ntn1\"}, {\"X\": \"miR-9\", \"V\": \"down-regulate\", \"Y\": \"Dcc nerve regeneration\"}]\n",
      "83 [{\"X\": \"Ultrafine carbon particles\", \"V\": \"down-regulate\", \"Y\": \"CYP1B1 expression in human monocytes .\"}]\n",
      "84 [{\"X\": \"Probiotics\", \"V\": \"down-regulate\", \"Y\": \"genes in Salmonella enterica serovar typhimurium pathogenicity islands 1 and 2 .\"}]\n",
      "85 [{\"X\": \"polyphenols\", \"V\": \"down-regulate\", \"Y\": \"stress protein tuberculosis\"}]\n",
      "86 [{\"X\": \"bmo-miR-0001\", \"V\": \"down-regulate\", \"Y\": \"expression gene\"}, {\"X\": \"bmo-miR-0015\", \"V\": \"down-regulate\", \"Y\": \"expression gene\"}]\n",
      "87 [{\"X\": \"suppressor cells\", \"V\": \"down-regulate\", \"Y\": \"expression\"}, {\"X\": \"suppressor cells\", \"V\": \"down-regulate\", \"Y\": \"CD8+ T cells\"}]\n",
      "88 [{\"X\": \"Glucocorticoids\", \"V\": \"down-regulate\", \"Y\": \"somatostatin receptors\"}]\n",
      "89 [{\"X\": \"Glucocorticoids\", \"V\": \"down-regulate\", \"Y\": \"the extracellular matrix proteins fibronectin , fibulin-1 and fibulin-2 in bone marrow stroma .\"}]\n",
      "90 [{\"X\": \"memory CTL\", \"V\": \"down-regulate\", \"Y\": \"molecule expression following virus clearance\"}]\n",
      "91 [{\"X\": \"MiR-126 restoration\", \"V\": \"down-regulate\", \"Y\": \"VEGF and inhibit the growth of lung cancer cell lines in vitro and in vivo .\"}]\n",
      "92 [{\"X\": \"miR-26a,miR-214\", \"V\": \"down-regulate\", \"Y\": \"expression of the PTEN gene in chronic lymphocytic leukemia , but not PTEN mutation or promoter methylation .\"}]\n",
      "93 [{\"X\": \"IL-12\", \"V\": \"down-regulate\", \"Y\": \"B cell migration manner\"}, {\"X\": \"IL-18\", \"V\": \"down-regulate\", \"Y\": \"B cell migration manner\"}]\n",
      "94 [{\"X\": \"VIP\", \"V\": \"down-regulate\", \"Y\": \"CXCL10\"}, {\"X\": \"PACAP\", \"V\": \"down-regulate\", \"Y\": \"CXCL10\"}]\n",
      "95 [{\"X\": \"concentrations\", \"V\": \"down-regulate\", \"Y\": \"agonist induced secretion\"}]\n",
      "96 [{\"X\": \"Flavonoids from Theobroma cacao\", \"V\": \"down-regulate\", \"Y\": \"inflammatory mediators .\"}]\n",
      "97 [{\"X\": \"Motif Ligand 1 cells\", \"V\": \"down-regulates\", \"Y\": \"the invasion of their metastatic melanoma cells .\"}]\n",
      "98 [{\"X\": \"Two independent mechanisms\", \"V\": \"down-regulate\", \"Y\": \"SecA ATPase activity\"}]\n",
      "99 [{\"X\": \"metabolites\", \"V\": \"down-regulate\", \"Y\": \"receptor gamma\"}]\n"
     ]
    }
   ],
   "source": [
    "result = {}\n",
    "\n",
    "# Summarize NP into a string\n",
    "def summarize_np(tree):\n",
    "    if tree.label() != 'NP':\n",
    "        return ''\n",
    "    \n",
    "    # NP -> N\n",
    "    if tree[0].label() == 'N':\n",
    "        return tree[0][0][0]\n",
    "    \n",
    "    # NP -> DT NP | J NP | R NP \n",
    "    if tree[0].label() in ('J', 'DT', 'R'):\n",
    "        return summarize_np(tree[1])\n",
    "    \n",
    "    # NP -> PART\n",
    "    if tree[0].label() == 'PART':\n",
    "        return tree[0][0][0]\n",
    "    \n",
    "    # NP -> WRONG_NP\n",
    "    # WRONG_NP: Originally, it should not be classified into NP\n",
    "    #           but it is classified as NP to mitigate not correctly tagged sentence\n",
    "    if tree[0].label() == 'WRONG_NP':\n",
    "        if tree[0][0].label() == 'JJ':\n",
    "            return tree[0][0][0]\n",
    "        \n",
    "        if tree[0][0].label() == 'R':\n",
    "            return tree[0][1][0][0]\n",
    "        \n",
    "        if tree[0][0].label() == 'NNP':\n",
    "            return tree[0][0][0] + ' ' + tree[0][1][0]\n",
    "    \n",
    "    if len(tree) < 2:\n",
    "        return ''\n",
    "    \n",
    "    # NP-> NP R | NP MOD\n",
    "    if tree[1].label() in ('MOD', 'R'):\n",
    "        return summarize_np(tree[0])\n",
    "\n",
    "    # NP -> NP C NP\n",
    "    if tree[1].label() == 'C':\n",
    "        return summarize_np(tree[0]) + ',' + summarize_np(tree[2])\n",
    "\n",
    "    # NP -> NP POS \n",
    "    if tree[1].label() == 'POS':\n",
    "        return summarize_np(tree[0]) + ' ' + tree[1][0]\n",
    "\n",
    "    # NP -> NP NP\n",
    "    if tree[1].label() == 'NP' and tree[0].label() == 'NP':\n",
    "        return summarize_np(tree[0]) + ' ' + summarize_np(tree[1])\n",
    "    \n",
    "    return ''\n",
    "\n",
    "# Summarize VP into a string\n",
    "def summarize_vp(tree):\n",
    "    # VP -> V\n",
    "    if tree[0].label() == 'V':\n",
    "        if isinstance(tree[0][0], str):\n",
    "            return tree[0][0]\n",
    "        \n",
    "        # V -> VB | VBD | VBP | VBZ\n",
    "        return tree[0][0][0]\n",
    "    \n",
    "    # VP -> R VP | MD VP\n",
    "    if tree[1].label() == 'VP':\n",
    "        return summarize_vp(tree[1])\n",
    "    \n",
    "    # VP -> VP C VP \n",
    "    if tree[1].label() == 'C':\n",
    "        return summarize_vp(tree[0]) + ',' + summarize_vp(tree[2])\n",
    "    \n",
    "    # VP -> VP PART | VP MOD\n",
    "    if tree[0].label() == 'VP':\n",
    "        return summarize_vp(tree[0])\n",
    "    \n",
    "    return ''\n",
    "\n",
    "# Summarize S into an array of { X: String, V: String, Z: String }\n",
    "def summarize_s(tree):\n",
    "    # S -> NP VP | NP VP NP \n",
    "    if tree[0].label() == 'NP':\n",
    "        np = summarize_np(tree[0])\n",
    "        vp = summarize_vp(tree[1])\n",
    "        np_2 = ''\n",
    "        \n",
    "        # S -> NP VP\n",
    "        if len(tree) <= 2:\n",
    "            # S -> NP (VP -> VP MOD)\n",
    "            if tree[1][1].label() == 'MOD':\n",
    "                np_2 = summarize_np(tree[1][1][1])\n",
    "        \n",
    "        # S -> NP VP NP\n",
    "        else:\n",
    "            np_2 = summarize_np(tree[2])\n",
    "        \n",
    "        if np_2 == '':\n",
    "            return []\n",
    "        \n",
    "        return [{ 'X': np, 'V': vp, 'Y': np_2 }]\n",
    "    \n",
    "    # S -> S \".\"\n",
    "    if isinstance(tree[1], str):\n",
    "        return summarize_s(tree[0])\n",
    "    \n",
    "    # S -> S C S\n",
    "    if tree[1].label() == 'C':\n",
    "        return summarize_s(tree[0]) + summarize_s(tree[2])\n",
    "    \n",
    "    # S -> VP NP\n",
    "    if tree[0].label() == 'VP':\n",
    "        return []\n",
    "    \n",
    "    # S -> S RB | S PART\n",
    "    if tree[0].label() == 'S':\n",
    "        return summarize_s(tree[0])\n",
    "    \n",
    "    # S -> WDT S | EX S | IN S\n",
    "    if tree[1].label() == 'S':\n",
    "        return summarize_s(tree[1])\n",
    "    \n",
    "    return []\n",
    "\n",
    "\n",
    "for idx, tree in parsed_sents.items():\n",
    "    values = summarize_s(tree)\n",
    "    \n",
    "    # if len(result[idx]) == 0:\n",
    "        # tree.pretty_print()\n",
    "    \n",
    "    new_values = []\n",
    "    # Change { X: 'a,b', V: 'bind', Y: 'c'} to\n",
    "    #        { X: 'a', V: 'bind', Y: 'c'}, { X: 'b', V: 'bind', Y: 'c'}\n",
    "    \n",
    "    for value in values:\n",
    "        xs = value['X'].split(',')\n",
    "        ys = value['Y'].split(',')\n",
    "        \n",
    "        for x in xs:\n",
    "            for y in ys:\n",
    "                new_values.append({'X': x, 'V': value['V'], 'Y': y})\n",
    "    \n",
    "    result[idx] = new_values\n",
    "\n",
    "\n",
    "# Parse NP to str\n",
    "def parse_np(subsent):\n",
    "    np_grammar, np_words = generate_sent_grammar(subsent)\n",
    "    \n",
    "    parser = ChartParser(CFG(Nonterminal('NP'), np_grammar))\n",
    "    tree = parser.parse_one(np_words)\n",
    "    \n",
    "    if tree is None:\n",
    "        return ' '.join(np_words)\n",
    "    \n",
    "    return summarize_np(tree)\n",
    "\n",
    "# If given sentence could not be parsed, split it into two parts using the verb.\n",
    "# Then, try to parse the parts as NP\n",
    "#     If it also fails, just return splitted.\n",
    "for idx, non_sent in non_sents.items():\n",
    "    verb_cands = list([ (word, tag) for word, tag in non_sent if tag == 'V' ])\n",
    "    if len(verb_cands) == 0:\n",
    "        verb_cands = list([ (word, tag) for word, tag in non_sent if tag.startswith('V') ])\n",
    "    \n",
    "    if len(verb_cands) == 0:\n",
    "        result[idx] = [{'X': '', 'V': '', 'Y': ''}]\n",
    "        continue\n",
    "    \n",
    "    verb = verb_cands[0]\n",
    "    verb_idx = non_sent.index(verb)\n",
    "    \n",
    "    # Split into two parts\n",
    "    sub_raw = non_sent[:verb_idx]\n",
    "    obj_raw = non_sent[verb_idx + 1:]\n",
    "    \n",
    "    sub = parse_np(sub_raw)\n",
    "    obj = parse_np(obj_raw)\n",
    "    \n",
    "    result[idx] = [{\n",
    "        'X': sub,\n",
    "        'V': verb[0],\n",
    "        'Y': obj\n",
    "    }]\n",
    "\n",
    "# Print extracted result\n",
    "import json\n",
    "with open('./output.txt', 'w', encoding='utf-8') as f:\n",
    "    for index, values in sorted(result.items(), key = lambda x: x[0]):\n",
    "        value_serialized = json.dumps(values)\n",
    "        \n",
    "        f.write(value_serialized + '\\n')\n",
    "        print(index, value_serialized)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR Poly- and perfluorinated compounds activate human pregnane X receptor.\n",
      "[{\"X\": \"Neurotransmitters\", \"V\": \"activate\", \"Y\": \"T-cells\"}]\n",
      "[{\"X\": \"Pericytes\", \"V\": \"activate\", \"Y\": \"complement\"}]\n",
      "[{\"X\": \"Molecules\", \"V\": \"activate\", \"Y\": \"different signaling pathways\"}]\n",
      "[{\"X\": \"Monoamines\", \"V\": \"activate\", \"Y\": \"neuropeptide signaling cascades\"}]\n",
      "[{\"X\": \"bacteria\", \"V\": \"activate\", \"Y\": \"type-I interferon production\"}]\n",
      "[{\"X\": \"Salient sounds\", \"V\": \"activate\", \"Y\": \"human visual cortex\"}]\n",
      "[{\"X\": \"Bisphenol A\", \"V\": \"activate\", \"Y\": \"human pregnane X receptor\"}, {\"X\": \"analogues\", \"V\": \"activate\", \"Y\": \"human pregnane X receptor\"}]\n",
      "[{\"X\": \"cells\", \"V\": \"activate\", \"Y\": \"T-cell responses\"}]\n",
      "[{\"X\": \"substhreshold sequences\", \"V\": \"activate\", \"Y\": \"retina\"}]\n",
      "[{\"X\": \"drugs\", \"V\": \"activate\", \"Y\": \"reward pathways\"}]\n",
      "[{\"X\": \"Cannabinoids\", \"V\": \"activate\", \"Y\": \"signaling\"}]\n",
      "[{\"X\": \"Integrins\", \"V\": \"activates\", \"Y\": \"nuclear transcription factor-kappaB\"}, {\"X\": \"cytokines\", \"V\": \"activates\", \"Y\": \"nuclear transcription factor-kappaB\"}]\n",
      "[{\"X\": \"Enhancer elements\", \"V\": \"activate\", \"Y\": \"3' splice site\"}]\n",
      "[{\"X\": \"actions\", \"V\": \"activate\", \"Y\": \"sulcus\"}, {\"X\": \"actions\", \"V\": \"activate\", \"Y\": \"fusiform gyrus\"}]\n",
      "[{\"X\": \"Muscarinic acetylcholine receptors\", \"V\": \"activate\", \"Y\": \"acetylcholinesterase gene promoter\"}]\n",
      "[{\"X\": \"peptide antibodies\", \"V\": \"activate\", \"Y\": \"complement\"}]\n",
      "[{\"X\": \"2-Aminopyridinium ions\", \"V\": \"activate\", \"Y\": \"nitroalkenes\"}]\n",
      "[{\"X\": \"antibodies\", \"V\": \"activate\", \"Y\": \"killing\"}]\n",
      "[{\"X\": \"compounds\", \"V\": \"activate\", \"Y\": \"TRPA1 ion channels\"}]\n",
      "[{\"X\": \"Bacterial polysaccharides\", \"V\": \"inhibit\", \"Y\": \"hyperglycemia\"}]\n",
      "[{\"X\": \"Iridoids\", \"V\": \"inhibit\", \"Y\": \"inflammation\"}, {\"X\": \"Iridoids\", \"V\": \"inhibit\", \"Y\": \"stress\"}]\n",
      "ERROR Aquatic pseudomonads inhibit Oomycete plant pathogens of glycine max.\n",
      "[{\"X\": \"Phenformin\", \"V\": \"inhibit\", \"Y\": \"growth\"}, {\"X\": \"metformin\", \"V\": \"inhibit\", \"Y\": \"growth\"}, {\"X\": \"Phenformin\", \"V\": \"inhibit\", \"Y\": \"migration\"}, {\"X\": \"metformin\", \"V\": \"inhibit\", \"Y\": \"migration\"}]\n",
      "[{\"X\": \"stem cells\", \"V\": \"inhibit\", \"Y\": \"T follicular helper cell\"}]\n",
      "[{\"X\": \"Bufadienolides\", \"V\": \"inhibit\", \"Y\": \"cell proliferation\"}]\n",
      "[{\"X\": \"Flavonoids\", \"V\": \"inhibit\", \"Y\": \"COX-1\"}, {\"X\": \"Flavonoids\", \"V\": \"inhibit\", \"Y\": \"COX-2 enzymes\"}, {\"X\": \"Flavonoids\", \"V\": \"inhibit\", \"Y\": \"cytokine/chemokine production\"}]\n",
      "[{\"X\": \"Procyanidins\", \"V\": \"inhibit\", \"Y\": \"tumor aniogenesis\"}]\n",
      "[{\"X\": \"hydrophobic aggregates\", \"V\": \"inhibit\", \"Y\": \"ice formation\"}]\n",
      "[{\"X\": \"molecular lipids\", \"V\": \"inhibit\", \"Y\": \"growth\"}]\n",
      "[{\"X\": \"Oxathiazolones\", \"V\": \"inhibit\", \"Y\": \"immunoproteasome\"}]\n",
      "[{\"X\": \"Largazole\", \"V\": \"inhibit\", \"Y\": \"ubiquitin activating enzyme\"}, {\"X\": \"derivatives\", \"V\": \"inhibit\", \"Y\": \"ubiquitin activating enzyme\"}]\n",
      "[{\"X\": \"Jun proteins\", \"V\": \"inhibit\", \"Y\": \"autophagy\"}]\n",
      "[{\"X\": \"Axl-148b chimeric aptamers\", \"V\": \"inhibit\", \"Y\": \"breast cancer\"}, {\"X\": \"Axl-148b chimeric aptamers\", \"V\": \"inhibit\", \"Y\": \"melanoma progression\"}]\n",
      "[{\"X\": \"Toll-like receptor 2 agonists\", \"V\": \"inhibit\", \"Y\": \"human fibrocyte differntiation\"}]\n",
      "[{\"X\": \"Ebola virus VP 24 proteins\", \"V\": \"inhibit\", \"Y\": \"interaction\"}]\n",
      "[{\"X\": \"Hawthorn extracts\", \"V\": \"inhibit\", \"Y\": \"LDL oxidation\"}]\n",
      "[{\"X\": \"membranes\", \"V\": \"inhibit\", \"Y\": \"prostaglandin\"}]\n",
      "[{\"X\": \"monocytes\", \"V\": \"inhibit\", \"Y\": \"multiplication\"}]\n",
      "[{\"X\": \"Heparin\", \"V\": \"inhibit\", \"Y\": \"zymosan\"}, {\"X\": \"Heparin\", \"V\": \"inhibit\", \"Y\": \"cobra venom factor activation\"}]\n",
      "[{\"X\": \"amphetamine\", \"V\": \"binds\", \"Y\": \"N-terminus\"}]\n",
      "[{\"X\": \"Type I inhibitors\", \"V\": \"bind\", \"Y\": \"active protein kinase conformation\"}]\n",
      "[{\"X\": \"TALE proteins\", \"V\": \"bind\", \"Y\": \"chromatin\"}]\n",
      "[{\"X\": \"\\u03b1-Synuclein\", \"V\": \"bind\", \"Y\": \"celluar membrane\"}, {\"X\": \"huntingtin exon 1 amyloid fibrils\", \"V\": \"bind\", \"Y\": \"cellular membrane\"}]\n",
      "[{\"X\": \"Catecholamines\", \"V\": \"bind\", \"Y\": \"enkephalins\"}, {\"X\": \"Catecholamines\", \"V\": \"bind\", \"Y\": \"morphiceptin\"}, {\"X\": \"Catecholamines\", \"V\": \"bind\", \"Y\": \"morphine\"}]\n",
      "[{\"X\": \"Acanthamoebae\", \"V\": \"bind\", \"Y\": \"rabbit corneal epithelium\"}]\n",
      "[{\"X\": \"Bordetella pertussis fimbriae\", \"V\": \"bind\", \"Y\": \"motocytes\"}]\n",
      "[{\"X\": \"pathogenic bacteria\", \"V\": \"bind\", \"Y\": \"carbohydrate sequence\"}]\n",
      "[{\"X\": \"TTAGGG repeat factors 1 and 2\", \"V\": \"bind\", \"Y\": \"subset\"}]\n",
      "[{\"X\": \"forms\", \"V\": \"bind\", \"Y\": \"bacteria\"}]\n",
      "[{\"X\": \"autoantibodies\", \"V\": \"bind\", \"Y\": \"T lymphocytes\"}]\n",
      "[{\"X\": \"CD3+ thymocytes\", \"V\": \"bind\", \"Y\": \"cells\"}]\n",
      "[{\"X\": \"Sulfoglucuronyl glycolipids\", \"V\": \"bind\", \"Y\": \"laminin\"}]\n",
      "[{\"X\": \"proteins\", \"V\": \"bind\", \"Y\": \"fibronectin\"}, {\"X\": \"proteins\", \"V\": \"bind\", \"Y\": \"laminin\"}, {\"X\": \"proteins\", \"V\": \"bind\", \"Y\": \"collagen\"}]\n",
      "[{\"X\": \"Microvesicles\", \"V\": \"bind\", \"Y\": \"soluble fibrinogen\"}]\n",
      "[{\"X\": \"Falciparum malaria-infected erythrocytes\", \"V\": \"bind\", \"Y\": \"cells\"}]\n",
      "[{\"X\": \"monemer\", \"V\": \"bind\", \"Y\": \"TLR-2\"}, {\"X\": \"peptides\", \"V\": \"bind\", \"Y\": \"TLR-2\"}]\n",
      "[{\"X\": \"scorpion toxins\", \"V\": \"bind\", \"Y\": \"receptor sites\"}]\n",
      "[{\"X\": \"fingers\", \"V\": \"bind\", \"Y\": \"DNA\"}]\n",
      "[{\"X\": \"Enzymes\", \"V\": \"bind\", \"Y\": \"subrates\"}]\n",
      "[{\"X\": \"Obesity\", \"V\": \"accelerate\", \"Y\": \"aging process\"}]\n",
      "[{\"X\": \"Antioxidants\", \"V\": \"accelerate\", \"Y\": \"lung cancer progression\"}]\n",
      "[{\"X\": \"Grooves\", \"V\": \"accelerate\", \"Y\": \"dew shedding\"}]\n",
      "[{\"X\": \"proton bunches\", \"V\": \"accelerate\", \"Y\": \"electrons\"}]\n",
      "[{\"X\": \"exosomes\", \"V\": \"accelerate\", \"Y\": \"osteogenesis\"}]\n",
      "[{\"X\": \"TFS\", \"V\": \"accelerate\", \"Y\": \"transcription\"}, {\"X\": \"Spt4/5\", \"V\": \"accelerate\", \"Y\": \"transcription\"}]\n",
      "[{\"X\": \"Eosinophils\", \"V\": \"accelerate\", \"Y\": \"pathogenesis\"}]\n",
      "[{\"X\": \"membranes\", \"V\": \"accelerate\", \"Y\": \"amyloid formation\"}]\n",
      "[{\"X\": \"Homocysteine\", \"V\": \"accelerate\", \"Y\": \"skin aging\"}]\n",
      "[{\"X\": \"NLRP3 inflammasome\", \"V\": \"accelerate\", \"Y\": \"pathology\"}, {\"X\": \"NLRP3 inflammasome\", \"V\": \"accelerate\", \"Y\": \"IL-1\\u03b2\"}]\n",
      "[{\"X\": \"glucosamine fibers\", \"V\": \"accelerate\", \"Y\": \"hemostatis\"}]\n",
      "[{\"X\": \"stem cells\", \"V\": \"accelerate\", \"Y\": \"wound healing\"}]\n",
      "[{\"X\": \"Cumulus cells\", \"V\": \"accelerate\", \"Y\": \"aging\"}]\n",
      "[{\"X\": \"Aluminium salts\", \"V\": \"accelerate\", \"Y\": \"peroxidation\"}]\n",
      "[{\"X\": \"foudnations\", \"V\": \"accelerate\", \"Y\": \"change\"}, {\"X\": \"foudnations\", \"V\": \"accelerate\", \"Y\": \"competence\"}]\n",
      "[{\"X\": \"Gangliosides\", \"V\": \"accelerate\", \"Y\": \"rat neonatal learning\"}, {\"X\": \"Gangliosides\", \"V\": \"accelerate\", \"Y\": \"levels\"}]\n",
      "[{\"X\": \"Semax\", \"V\": \"accelerate\", \"Y\": \"healing\"}, {\"X\": \"glyproline peptides\", \"V\": \"accelerate\", \"Y\": \"healing\"}]\n",
      "[{\"X\": \"engines\", \"V\": \"accelerate\", \"Y\": \"aging\"}, {\"X\": \"engines\", \"V\": \"accelerate\", \"Y\": \"disease\"}]\n",
      "[{\"X\": \"MIF-1\", \"V\": \"accelerate\", \"Y\": \"neuromotor\"}, {\"X\": \"MIF-1\", \"V\": \"accelerate\", \"Y\": \"EEG\"}, {\"X\": \"MIF-1\", \"V\": \"accelerate\", \"Y\": \"development\"}]\n",
      "[{\"X\": \"Distributed representations\", \"V\": \"accelerate\", \"Y\": \"evolution\"}]\n",
      "[{\"X\": \"Azelogycine\", \"V\": \"down-regulate\", \"Y\": \"mediators\"}, {\"X\": \"nicotinamide\", \"V\": \"down-regulate\", \"Y\": \"mediators\"}]\n",
      "[{\"X\": \"miR-214\", \"V\": \"down-regulate\", \"Y\": \"Necl-2/CADM1\"}, {\"X\": \"hypoxia\", \"V\": \"down-regulate\", \"Y\": \"Necl-2/CADM1\"}]\n",
      "[{\"X\": \"microRNAs let-7\", \"V\": \"down-regulate\", \"Y\": \"axon-guidance genes\"}, {\"X\": \"mir-9\", \"V\": \"down-regulate\", \"Y\": \"axon-guidance genes\"}]\n",
      "[{\"X\": \"carbon particles\", \"V\": \"down-regulate\", \"Y\": \"CYP1B1 expression\"}]\n",
      "[{\"X\": \"Probiotics\", \"V\": \"down-regulate\", \"Y\": \"genes\"}]\n",
      "[{\"X\": \"polyphenols\", \"V\": \"down-regulate\", \"Y\": \"stress protein\"}]\n",
      "[{\"X\": \"bmo-miR-0001\", \"V\": \"down-regulate\", \"Y\": \"expression\"}, {\"X\": \"bmo-miR-0015\", \"V\": \"down-regulate\", \"Y\": \"expression\"}]\n",
      "[{\"X\": \"suppressor cells\", \"V\": \"down-regulate\", \"Y\": \"L-selectin expression\"}]\n",
      "[{\"X\": \"Glucocorticoids\", \"V\": \"down-regulate\", \"Y\": \"somatostatin receptors\"}]\n",
      "[{\"X\": \"Glucocorticoids\", \"V\": \"down-regulate\", \"Y\": \"matrix proteins fibronectin\"}, {\"X\": \"Glucocorticoids\", \"V\": \"down-regulate\", \"Y\": \"matrix proteins fibulin-1\"}, {\"X\": \"Glucocorticoids\", \"V\": \"down-regulate\", \"Y\": \"matrix proteins fibulin-2\"}]\n",
      "[{\"X\": \"memory CTL\", \"V\": \"down-regulate\", \"Y\": \"molecule expression\"}]\n",
      "[{\"X\": \"MiR-126 restoration\", \"V\": \"down-regulate\", \"Y\": \"VEGF\"}]\n",
      "[{\"X\": \"miR-26a\", \"V\": \"down-regulate\", \"Y\": \"expression\"}, {\"X\": \"miR-214\", \"V\": \"down-regulate\", \"Y\": \"expression\"}]\n",
      "[{\"X\": \"IL-12\", \"V\": \"down-regulate\", \"Y\": \"B cell migration\"}, {\"X\": \"IL-18\", \"V\": \"down-regulate\", \"Y\": \"B cell migration\"}]\n",
      "[{\"X\": \"VIP\", \"V\": \"down-regulate\", \"Y\": \"CXCL10\"}, {\"X\": \"PACAP\", \"V\": \"down-regulate\", \"Y\": \"CXCL10\"}]\n",
      "[{\"X\": \"concentrations\", \"V\": \"down-regulate\", \"Y\": \"secretion\"}]\n",
      "[{\"X\": \"Flavonoids\", \"V\": \"down-regulate\", \"Y\": \"mediators\"}]\n",
      "[{\"X\": \"C-X-C Motif Ligand 1\", \"V\": \"down-regulates\", \"Y\": \"invasion\"}]\n",
      "[{\"X\": \"mechanisms\", \"V\": \"down-regulate\", \"Y\": \"SecA ATPase activity\"}]\n",
      "[{\"X\": \"metabolites\", \"V\": \"down-regulate\", \"Y\": \"receptor gamma\"}]\n",
      "TP       : 1\n",
      "FP       : 135\n",
      "FN       : 134\n",
      "Precision: 0.0074\n",
      "Recall   : 0.0074\n",
      "F Score  : 0.0074\n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "with open('dataset/dataset_annotations.csv', encoding='utf-8') as csvfile:\n",
    "    reader = csv.reader(csvfile)\n",
    "    rows = list(reader)\n",
    "\n",
    "true_positive = 0\n",
    "false_positive = 0\n",
    "false_negative = 0\n",
    "\n",
    "for idx, (annotations, _) in enumerate(rows[USING_PORTION]):\n",
    "    groundtruth = set()\n",
    "    for triplet in json.loads(annotations):\n",
    "        groundtruth.add(\n",
    "            (triplet['X'], triplet['V'], triplet['Y'])\n",
    "        )\n",
    "\n",
    "    predicted = set()\n",
    "    for triplet in result[idx]:\n",
    "        predicted.add(\n",
    "            (triplet['X'], triplet['V'], triplet['Y'])\n",
    "        )\n",
    "    \n",
    "    true_positive += len(groundtruth & predicted)\n",
    "    false_positive += len(predicted - groundtruth)\n",
    "    false_negative += len(groundtruth - predicted)\n",
    "\n",
    "precision = true_positive / (true_positive + false_positive)\n",
    "recall = true_positive / (true_positive + false_negative)\n",
    "\n",
    "if precision + recall == 0:\n",
    "    f_score = 0\n",
    "\n",
    "else:\n",
    "    f_score = 2 * (precision * recall) / (precision + recall)\n",
    "\n",
    "\n",
    "print(\"TP       : %d\"   % true_positive)\n",
    "print(\"FP       : %d\"   % false_positive)\n",
    "print(\"FN       : %d\"   % false_negative)\n",
    "print(\"Precision: %.4f\" % precision)\n",
    "print(\"Recall   : %.4f\" % recall)\n",
    "print(\"F Score  : %.4f\" % f_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
